workers/unity/Packages/com.improbable.gdk.querybasedinteresthelper/InterestBuilder.cs

```diff
@@ -0,0 +1,177 @@
+using System.Collections.Generic;
+using Improbable.Gdk.Core;
+
+namespace Improbable.Gdk.QueryBasedInterest
+{
+
+    /// <summary>
+    ///     Utility class to help construct Interest component snapshots.
+    /// </summary>
+    public class InterestBuilder
+    {
+        private readonly Dictionary<uint, ComponentInterest> interest;
+
+        private InterestBuilder(Dictionary<uint, ComponentInterest> interest)
+        {
+            this.interest = interest ?? new Dictionary<uint, ComponentInterest>();
+        }
+
+        /// <summary>
+        ///     Creates a new InterestBuilder object.
+        /// </summary>
+        /// <returns>
+        ///     A new InterestBuilder object.
+        /// </returns>
+        public static InterestBuilder Begin()
+        {
+            return new InterestBuilder(null);
+        }
+
+        /// <summary>
+        ///     Creates a new InterestBuilder object from the data of an existing Interest component.
+        /// </summary>
+        /// <param name="interest">
+        ///     The data of an existing Interest component.
+        /// </param>
+        /// <returns>
+        ///     A new InterestBuilder object.
+        /// </returns>
+        public static InterestBuilder Modify(Dictionary<uint, ComponentInterest> interest)
```

Comment:
This is kind of like a copy constructor. I'm not a hugeee fan of the name `Modify`.

Also note that the semantics here mean that you mutate the dictionary you pass into the constructor. That may not be what the API suggests (e.g. - a base set of interests which you append things to).

You'll need to do a deep copy of the dictionary to copy it

Reply:
The intention here is to mutate the dictionary you pass in. A potential use case for that would be adding/removing a query depending on some external factor when the game is running. In which case you'd do something like:

```
InterestBuilder.Modify(InterestWriter.Data.ComponentInterest)
    .ReplaceQueries<Position.Component>(...)
    .GetInterest();
```

Reply:
thinking aloud - since it's modifying it directly the `GetInterest` method wouldn't be needed in the current implementation

however - is this the best way to give this sort of functionality to users?

Reply:
Give the signature of the API I would expect this would seed my initial interest with the dictionary and then allow me to modify it, w/o modifying the original dictionary. 

Maybe an `InterestBuilder.From(GeneratedComponentName component)` if the intent is to mutate an existing component?

Reply:
so now there's 

`.Begin()` - fresh new interest builder

`.From(...)` - deep copy a given Interest component/dictionary and modify the copy

`.Mutate(...)` - modify the given Interest component/dictionary directly